opm/test_util/summaryIntegrationTest.cpp

```diff
@@ -0,0 +1,352 @@
+	    if(volume.error == 0){
+		std::cout << "For keyword " << keyword << " the total production volume is 0" << std::endl; 
+	    }
+	    else{
+		std::cout << "For keyword " << keyword << " the total production volume is "<< volume.total << ", the error volume is " << volume.error << " the error ratio is " << volume.error/volume.total << std::endl;
+	    }
+	}  
+  	checkForKeyword(timeVec1, timeVec2, keyword);
+	return;
+    } 
+    OPM_THROW(std::invalid_argument, "The keyword used is not common for the two files.");
+  
+}
+
+
+void IntegrationTest::checkForKeyword(std::vector<double> timeVec1, std::vector<double> timeVec2, const char* keyword){
+    std::vector<double> dataVec1, dataVec2;
+  
+    getDataVecs(dataVec1,dataVec2,keyword); 
+    chooseReference(timeVec1, timeVec2,dataVec1,dataVec2);
+ 
+    if(allowSpikes){
+	checkWithSpikes(keyword);
+    }  
+
+    if(findVolumeError ||oneOfTheMainVariables ){
+	volumeErrorCheck(keyword);
+    }
+}
+
+
+
+
+
+int IntegrationTest::checkDeviation(Deviation deviation){
+    double absTol = getAbsTolerance();
+    double relTol = getRelTolerance();
+
+    if (deviation.rel> relTol && deviation.abs > absTol){
+	return 1;
+    }
+    return 0;
+}
+
+
+void IntegrationTest::findGreatestErrorRatio(WellProductionVolume volume, double &greatestRatio, const char* currentKeyword, std::string &keywordWithGreatestErrorRatio){
+    if (volume.total != 0 && volume.total - volume.error > getAbsTolerance()){
+	if(volume.error/volume.total > greatestRatio){
+	    greatestRatio = volume.error/volume.total; 
+	    std::string currentKeywordStr(currentKeyword);
+	    keywordWithGreatestErrorRatio = currentKeywordStr; 
+	}
+    }
+}
+
+
+void IntegrationTest::volumeErrorCheck(const char* keyword){
+    std::string keywordString(keyword);
+    std::string firstFour = keywordString.substr(0,4);
+    std::string fifth = keywordString.substr(4,1); //To make sure we do not include history vectors.
+   
+    if (mainVariable.empty()){
+	updateVolumeError(keyword);
+    }
+    
+    if(firstFour ==  mainVariable && fifth!="H"){
+	if(mainVariable == "WOPR"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WOPR.total += result.total; 
+	    WOPR.error += result.error; 
+	}
+	if(mainVariable == "WWPR"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WWPR.total += result.total; 
+	    WWPR.error += result.error; 
+	}
+
+	if(mainVariable == "WGPR"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WGPR.total += result.total; 
+	    WGPR.error += result.error; 
+	}
+
+	if(mainVariable == "WBHP"){
+	    WellProductionVolume result =  getWellProductionVolume(keyword);
+	    WBHP.total += result.total; 
+	    WBHP.error += result.error; 
+	}	
+    }
+}
+
+
+void IntegrationTest::updateVolumeError(const char* keyword){
+    std::string keywordString(keyword);
```

Comment:
> You do reinvent quite a lot of ert code - a bit unnecceassary in my opinion.

Not in this case I believe. ERT does not provide the integral of the WBHP curves to my knowledge, so this would need to be implemented somehow. When the same routine works for the rate curves, using calls to some code outside OPM seems only to obfuscate the code. Still, for such a simple operation, there should be no need to use an external routine which limited to one file format.

To appreciate the approach here, I think it is important to look at the intent. The intent of the implementation under discussion is to obtain some notion of when two simulations are close. Typically comparing simulations from two different simulators. In this case, for the summary part, we need to focus only on the main solution variables from the well equations. Hence, only production rates and BHP is checked. In this case we need not even have totals available. Checking that totals are calculated correctly is already catered to in the regression tests, here we want the test to work even when reported totals are wrong or not written out at all.

> Observe that iff you insist on determing total volumes by integrating rates the correct representation of the function R (t) is as a piecewise constant function,

Yes, so the integral should boil down to a simple Riemann sum.

> where the constant value is given as the rate at the upper limit of the interval.

This sounds wrong, but maybe I misunderstand here.

> Integrating WBHP does not give a volume but rather something like momentum per area .

Yes, so a renaming can be wise here. For instance removing word `volume` or replacing it by `integral` should make it clear. In either case, we do need to check relative errors, so the integral is necessary.